wayland-client/src/conn.rs

```diff
@@ -162,16 +162,12 @@ impl Connection {
         loop {
             self.backend.flush()?;
 
-            // first, prepare the read
-            let guard = self.backend.prepare_read()?;
-
-            // If another thread processed events prior to prepare_read, we might already be done.
-            if done.done.load(Ordering::Relaxed) {
-                break;
+            if let Some(guard) = self.backend.prepare_read() {
+                dispatched += blocking_read(guard)?;
+            } else {
+                dispatched += self.backend.dispatch_inner_queue()?;
```

Comment:
Hm, but you could have a blocking dispatch of the queue here as well? Non-blocking roundtrip doesn't sound like it'll work reliably or I don't understand something?

Reply:
This is in a loop, so it'll just try again at the next iteration.

Reply:
Ah, yeah, sure, I just know that there's mo efficient API to do the exact same thing, but I think it's not an option for us. The `wl_display_dispatch_queue` https://wayland.freedesktop.org/docs/html/apb.html#Client-classwl__display_1ae027b09801474ac7c6b0f1ef25ff6e17

Reply:
That function in wayland-client does mostly exactly the same thing as `wl_display_roundtrip`, but implemented in terms of wayland-backend's API, so I don't really get your point?

Reply:
`wl_display_roundtrip` uses `poll` under the hood to not do the loop.

Reply:
What... ?

`wl_display_roundtrip` does have a loop: https://gitlab.freedesktop.org/wayland/wayland/-/blob/main/src/wayland-client.c#L1402

and `blocking_read` uses `poll` internally: https://github.com/Smithay/wayland-rs/blob/master/wayland-client/src/conn.rs#L224-L245

Reply:
@elinorbgr if you look closely what their code is and what we have, they use blocking dispatch. However we use it only in `read_guard` branch, and the other branch is using regular non-blocking dispatch.

So while they have a `loop`, the `wait`, but we `busy loop` from the `else`. The blocking read is fine, the `else` is a busy loop. `dispatch_inner_queue` is a non_blocking API from what I can, see, but libwayland always using blocking API, with `dispatch_queue`.

Reply:
After reading the libwayland impl it's pretty much the same, when their docs are not really clear what exactly blocks and how it works.